kernel/src/net/socket/netlink/mod.rs

```diff
@@ -38,14 +38,97 @@
 //!
 
 mod addr;
+mod common;
+mod kobject_uevent;
 mod message;
+mod options;
 mod route;
 mod table;
 
+use addr::NetlinkProtocolId;
 pub use addr::{GroupIdSet, NetlinkSocketAddr};
+pub use kobject_uevent::NetlinkUeventSocket;
+pub use options::{AddMembership, DropMembership};
 pub use route::NetlinkRouteSocket;
 pub use table::{is_valid_protocol, StandardNetlinkProtocol};
 
+use crate::prelude::*;
+
 pub(in crate::net) fn init() {
     table::init();
 }
+
+#[derive(Clone)]
+enum AnyNetlinkSocket {
+    Route(Weak<NetlinkRouteSocket>),
+    Uevent(Weak<NetlinkUeventSocket>),
+}
+
+impl AnyNetlinkSocket {
+    const fn protocol(&self) -> NetlinkProtocolId {
+        match self {
+            Self::Route(_) => StandardNetlinkProtocol::ROUTE as _,
+            Self::Uevent(_) => StandardNetlinkProtocol::KOBJECT_UEVENT as _,
+        }
+    }
+
+    fn enqueue_unicast_message(&self, message: AnyUnicastMessage) -> Result<()> {
+        assert_eq!(self.protocol(), message.protocol());
+        match self {
+            Self::Route(socket) => {
+                let Some(socket) = socket.upgrade() else {
+                    return Ok(());
+                };
+                let AnyUnicastMessage::Route(message) = message;
+                socket.enqueue_message(message)
+            }
+            Self::Uevent(_) => todo!(),
+        }
+    }
+
+    fn enqueue_mutlicast_message(&self, message: AnyMulticastMessage) -> Result<()> {
+        assert_eq!(self.protocol(), message.protocol());
+        match self {
+            Self::Route(_) => todo!(),
```

Comment:
Is this reachable? What's the Linux behavior if it's reachable?

Reply:
Thanks for the explanation.

Indeed, the code here is confusing because we don't have `AnyUnicastMessage::Route`, so this is not a `todo!()`. It should be `unreachable!()`.

Hopefully, https://github.com/asterinas/asterinas/pull/2109#discussion_r2125697677 will resolve the confusion by having the Rust type system track the message type.